### PR TITLE
PerfMsgr: A tool to benchmark messenger module

### PR DESCRIPTION
**Pull request by yuyuyu101**
Just like iperf usage for basic network benchmark
#./ceph_perf_msgr_server 172.16.30.181:10001 0
using ms-type async
       bind ip:port 172.16.30.181:10001
       thinktime(us) 0

#./ceph_perf_msgr_client 172.16.30.181:10001 1 32 10000 10 4096 
using ms-type async
       server ip:port 172.16.30.181:10001
       numjobs 1
       concurrency 32
       ios 10000
       thinktime(us) 10
       message data bytes 4096

 Total op 10000 run time 852670us.

This tool can help find the bottleneck or time consuming within messenger module

### CODE REVIEW

File: src/test/msgr/perf_msgr_client.cc [hunk 1]
```diff
@@ -0,0 +1,205 @@
...
+}
+
+int main(int argc, char **argv)
+{
+  vector<const char*> args;
+  argv_to_vec(argc, (const char **)argv, args);
+
+  global_init(NULL, args, CEPH_ENTITY_TYPE_CLIENT, CODE_ENVIRONMENT_UTILITY, 0);
+  common_init_finish(g_ceph_context);
+  g_ceph_context->_conf->apply_changes(NULL);
```
Comment (gregsfortytwo) [inline src/test/msgr/perf_msgr_client.cc]: You shouldn't need this call right after doing the init setup?

**Comment by gregsfortytwo** [comment]
It looks like your description got cut short, can you complete it? ("This tool can help find the")

**Comment by gregsfortytwo** [comment]
What exactly are you trying to measure with this tool? I notice that it only works on fast_dispatch messages, and it lets you specify a time to fake for message processing in that pipeline, but there's no control over how long the back-end code takes, it just replies immediately.

**Comment by yuyuyu101** [comment]
@gregsfortytwo yeah, I want to implement a tool which simply do messenger module works especially for CEPH_OSD_OP message. So when facing *very fast* storage device backend, I guess we can easily find the bottleneck for specified messenger impl and do some code optimization. Because a real backend like ceph cluster which is affected by many things always can' t tell the realistic performance data which help to dig bottleneck.

**Comment by yuyuyu101** [comment]
I dont add any sleep time to backend-end code is that we only want to perf messenger module. We need to make backend-end code as fast as possible

**Comment by gregsfortytwo** [comment]
But what's the point of making fast dispatch more or less expensive if we want to focus on the messenger? The cost of fast dispatch is in the Dispatcher's routines, not the Messenger's.

And while I understand the point of simulating various relative costs, leaving out any mechanism for injecting Dispatcher-internal CPU costs seems like it's important enough that we need that as well as the fast dispatch cost simulator to get anything like meaningful data.

**Comment by yuyuyu101** [comment]
@gregsfortytwo hmm, it maybe looked a little fool, my thought is that current msgr impl all did fast_dispatch inline which means receiver thread will be suck into fast dispatcher logic for a while. So compared async msgr to simple msgr,  if no "think time" in perf_msgr, async msgr will behavior happier than simple msgr. Because why async msgr is bad than simple in heavy benchmark is that each async thread consume much time in fast_dispatch. In other word, if we have 64 very active connections, 64 pipe threads can do it happily but less async thread(maybe 2,4) will spend more time in fast dispatch instead of reading socket and process message.

That's why I inject sleep time to fast dispatch not to backend thread. I'm looking for a way to solve the async performance problem related to fast dispatch or decode_message logic.

**Comment by gregsfortytwo** [comment]
Ah, now I understand. That makes a lot of sense!

This looks good to me. Can you write up some docs for this tool (in doc/dev and/or the help text) describing its use case with some example configurations for useful tests?

**Comment by markhpc** [comment]
FWIW, I've wanted something like this for years.  Awesome work Haomai! :)

**Pull request merged**